*******************************
Interpretation logic and output
*******************************

Interpreted AMRFinderplus output
================================

When given an AMRFinderPlus input file, and an organism, AMRrules will apply the most relevant organism-specific rule to each AMRFinderPlus hit. Below is an abbreviated example of the interpreted output file generated by AMRrules. (More examples can be found in **TESTS TBD**).

.. image:: images/interpreted_genotype_report.png
   :alt: Abbreviated example of an interpreted AMRFP table
   :align: center


When parsing the AMRFinderPlus input file, AMRrules will first identify what ``variation type`` each hit belongs to, the ``gene`` found, and if there is a mutation in the gene, this mutation will be converted to the :ref:`AMRrules syntax <syntax>`.

The table below shows the logic of how AMRrules maps AMRFinderPlus outputs to the different variation types:

================================ ================================ =====================================================
Variation type                   AMRFP Method                     Additional logic
================================ ================================ =====================================================
Gene presence detected           EXACT, ALLELE, BLAST
Protein variant detected         POINTX, POINTP
Nucleotide variant detected      POINTN
Promoter variant detected        POINTN                           ``-`` symbol present before the nucleotide position
Inactivating mutation detected   INTERNAL_STOP, PARTIAL, POINTX   If POINTX, Subtype must be POINT_DISRUPT
================================ ================================ =====================================================
                           

The parsed ``variation type``, ``gene``, and ``mutation`` columns are added to the interpreted output file, next to the original AMRFinderPlus columns.

Rule matching
^^^^^^^^^^^^^

When matching a hit with a rule, the following logic applies:

1. AMRrules will select all rules matching the ``variation type``.
2. AMRrules will filter all possible rules to check if the ``nodeID`` from the rule matches to ``Hierarchy node``. If not, it will check the parent nodes of ``Hierarchy node`` to look for a matching ``nodeID``.
3. If no matching ``nodeID`` is found, AMRrules will then check for matching accessions in the following order:

   a. ``nucleotide accession``
   b. ``protein accession``
   c. ``HMM accession``

For each matching rule, AMRrules appends the following information from the rules file to the AMRFinderPlus hit:

* ruleID
* gene context
* drug
* drug class
* phenotype
* clinical category
* evidence grade
* version
* organism

To add the full rule annotation information, set ``--annot-opts full``, which will add the following additional fields from the rule file:

* breakpoint
* breakpoint standard
* breakpoint condition
* evidence code
* evidence limitations
* PMID
* rule curation note

The resulting output file is stored as ``<output_prefix>_interpreted.tsv``.


Genome summary report
=====================

In addition to the interpreted AMRFinderPlus output file, AMRrules will also generate a genome summary report, stored as ``<output_prefix>_summary.tsv``. This file summarises the resistance per drug/drug class, based on the matched rules.

The following columns are included:

=================== ====================================================================================================
  Column              Explanation                                                                                                              
=================== ====================================================================================================
  sample              Sample ID                                                                                                                
  drug                Drug, as per CARD ARO. AMRFinderPlus drugs will be converted to CARD ARO format by AMRrules                              
  drug class          Drug class, as per CARD ARO. AMRFinderPlus drugs will be converted to CARD ARO format by AMRrules                       
  clinical category   S/I/R. Highest level of resistance based on markers that match this drug/drug class                                  
  phenotype           Wildtype or nonwildtype. Highest level based on markers that match this drug/drug class                                  
  evidence grade      Very low/low/moderate/high. Highest grade of evidence for markers that match this drug/drug class                       
  markers (non-S)     Markers with a matching rule that are predicted to confer an I or R phenotype, separated by ``;``
  markers (no rule)   Markers with no rule, separated by ``;``                                                                 
  markers (S)         Markers with a rule that are predicted to confer an S phenotype, separated by ``;``                      
  ruleIDs             List of ruleIDs that apply to this drug/drug class, separated by ``;``                                                     
  combo rules         All combination rules that apply to this drug/drug class                                                                 
  organism            Organism these markers apply to                                                                                          
=================== ====================================================================================================

Markers are formatted as ``gene:mutation``, where the mutation is formatted using :ref:`AMRrules syntax <syntax>`. Genes which are inactivated are depicted as ``gene:-``, where the ``-`` indicates an inactivating mutation.

.. note::

   AMRFinderPlus hits detected using POINT_DISRUPT contain detailed information about the disruption, which is formatted using HGVS syntax. For simplicity, by default AMRrules **will not show** the full mutation after the gene when listing these markers in the genome summary. Rather, they will be formatted as ``gene:-`` as per other inactivating mutations. If you wish to show the full mutation detected by AMRFinderPlus, this can be turned on by providing the option ``--full-disrupt`` to the AMRrules call.

Markers that are core genes can be optionally flagged in the genome summary output by turning on ``--flag-core``. eg the core blaSHV gene in *K. pneumoniae* will now be in the ``markers (non-S)`` column as ``blaSHV-11 (core)`` under ampicillin.


Handling unmatched hits
^^^^^^^^^^^^^^^^^^^^^^^^^^

By default, if no organism-specific rule is found for a hit, AMRrules will apply a prediction based on the setting of ``--no-rule-interpretation``. By default, this is set to ``nwtR``: any unmatched hit will be given the phenotype ``nonwildtype`` and the clinical category of resistance will be set to ``R``. Users can change this to be ``nwtS``, if they do not wish to call unmatched hits as resistant.

In all cases, hits with no rules have an evidence grade of ``very low`` for the purposes of the summary report.
