*******************************
Interpretation logic and output
*******************************

Interpreted AMRFinderplus output
================================

When given an AMRFinderPlus input file, and an organism, AMRrules will apply the most relevant organism-specific rule to each AMRFinderPlus hit. Below is an abbreviated example of the interpreted output file generated by AMRrules. (More examples can be found in **TESTS TBD**).

.. image:: images/interpreted_genotype_report.png
   :alt: Abbreviated example of an interpreted AMRFP table
   :align: center


When parsing the AMRFinderPlus input file, AMRrules will first identify what ``variation type`` each hit belongs to, the ``gene`` found, and if there is a mutation in the gene, this mutation will be converted to the AMRrules syntax **(link to syntax info here)**.

The table below shows the logic of how AMRrules maps AMRFinderPlus outputs to the different variation types:

================================ ================================ =====================================================
Variation type                   AMRFP Method                     Additional logic
================================ ================================ =====================================================
Gene presence detected           EXACT, ALLELE, BLAST
Protein variant detected         POINTX, POINTP
Nucleotide variant detected      POINTN
Promoter variant detected        POINTN                           ``-`` symbol present before the nucleotide position
Inactivating mutation detected   INTERNAL_STOP, PARTIAL, POINTX   If POINTX, Subtype must be POINT_DISRUPT
================================ ================================ =====================================================
                           

The parsed ``variation type``, ``gene``, and ``mutation`` columns are added to the interpreted output file, next to the original AMRFinderPlus columns.

Rule matching
^^^^^^^^^^^^^

When matching a hit with a rule, the following logic applies:

1. AMRrules will select all rules matching the ``variation type``.
2. AMRrules will filter all possible rules to check if the ``nodeID`` from the rule matches to ``Hierarchy node``. If not, it will check the parent nodes of ``Hierarchy node`` to look for a matching ``nodeID``.
3. If no matching ``nodeID`` is found, AMRrules will then check for matching accessions in the following order:

   a. ``nucleotide accession``
   b. ``protein accession``
   c. ``HMM accession``

For each matching rule, AMRrules appends the following information from the rules file to the AMRFinderPlus hit:

* ruleID
* gene context
* drug
* drug class
* phenotype
* clinical category
* evidence grade
* version
* organism

To add the full rule annotation information, set ``--annot_opts full``, which will add the following additional fields from the rule file:

* breakpoint
* breakpoint standard
* breakpoint condition
* evidence code
* evidence limitations
* PMID
* rule curation note

The resulting output file is stored as ``<output_prefix>_interpreted.tsv``.


Genome summary output
=====================



By default, if no organism-specific rule is found for a hit, AMRrules will apply a generic call based on the setting for ``--no-rule-interpretation``. By default, this is set to ``nwtR``: any unmatched hit will be given the phenotype ``nonwildtype`` and the clinical category of resistance will be set to ``R``. Users can change this to be ``nwtS``, if they do not wish to call unmatched hits as resistant.
